infrastructure/iac/iac.go

```diff
@@ -90,16 +98,46 @@ func (iac *Scanner) Scan(ctx context.Context, path string, _ string) (issues []s
 	} else {
 		workspacePath = filepath.Dir(uri.PathFromUri(documentURI))
 	}
+	iac.mutex.Lock()
+	i := scanCount
+	previousScan, wasFound := iac.runningScans[documentURI]
+	if wasFound && !previousScan.IsDone() { // If there's already a scan for the current workdir, we want to cancel it and restart it
+		previousScan.CancelScan()
+	}
+	newScan := scans.NewScanProgress()
+	go newScan.Listen(cancel, i)
+	defer func() {
+		iac.mutex.Lock()
+		log.Debug().Msgf("Scan %v is done", i)
+		newScan.SetDone() // Calling SetDone is safe even after cancellation
+		iac.mutex.Unlock()
+	}()
+	scanCount++
+	iac.runningScans[documentURI] = newScan
+	iac.mutex.Unlock()
 
 	scanResults, err := iac.doScan(ctx, documentURI, workspacePath)
 	p.Report(80)
 	if err != nil {
-		if err != context.Canceled {
+		noCancellation := ctx.Err() == nil
+		if noCancellation { // Only reports errors that are not intentional cancellations
 			iac.errorReporter.CaptureError(err)
+		} else { // If the scan was cancelled, return empty results
+			return issues
 		}
 	}
+
+	issues = iac.retrieveIssues(ctx, scanResults, issues, workspacePath, err)
+	return issues
+}
+
+func (iac *Scanner) retrieveIssues(ctx context.Context, scanResults []iacScanResult, issues []snyk.Issue, workspacePath string, err error) []snyk.Issue {
 	if len(scanResults) > 0 {
 		for _, s := range scanResults {
+			if ctx.Err() != nil {
```

Comment:
I'm wondering if we should really cancel here, if we have results - the parsing only takes milliseconds, so we could just return the result and be done ;).